src/js/reducers/conversation-reducer.js

```diff
@@ -36,22 +34,80 @@ const addMessage = (messages, message) => {
     return sortMessages([...messages, message]);
 };
 
+
+const matchMessage = (message, queryProps) => Object.keys(queryProps).every((key) => message[key] === queryProps[key]);
+
+const replaceMessage = (messages, query, newMessage) => {
+    const existingMessage = messages.find((message) => matchMessage(message, query));
+    if (!existingMessage) {
+        return messages;
+    }
+
+    const index = messages.indexOf(existingMessage);
+
+    // get rid of duplicate values
```

Comment:
the duplication removal logical should happen where you merge the two arrays, not where you replace a message